staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+		//    response is still in use
+		//
+		// when we enable write timeout on the server we see the following:
+		// server:
+		//  - write timeout happens after 50ms, the handler writes the
+		//    payload partially only.
+		//  - Write or FlushError return an 'i/o timeout' error
+		//  - the handler returns immediately after the write timeout
+		//
+		// client:
+		//  - client will finish reading the partial payload from the
+		//    response body, and then it will see a stream reset error
+		b := make([]byte, 125)
+		now := time.Now()
+		var read, iteration int
+		for {
+			iteration++
+			n, err := io.ReadFull(resp.Body, b)
+			if err != nil {
+				clientErrCh <- err
+				break
+			}
+			read += n
+			time.Sleep(time.Millisecond)
+			if iteration%1000 == 0 {
+				t.Logf("elapsed=%s, read=%dkB", time.Since(now), read/1000)
+			}
+		}
+		t.Logf("elapsed=%s, read=%dKB", time.Since(now), read/1000)
+	}()
+
+	handlerWait, clientWait := time.After(wait.ForeverTestTimeout), time.After(wait.ForeverTestTimeout)
+	err := handlerErrCh.wait(t, handlerWait)
+	wantError{err: os.ErrDeadlineExceeded}.verify(t, err)
+
+	err = clientErrCh.wait(t, clientWait)
+	wantContains{contains: "stream error: stream ID 1; INTERNAL_ERROR; received from peer"}.verify(t, err)
+}
+
+func sendWithTrace(t *testing.T, client *http.Client, url string, f func(*testing.T, httptrace.GotConnInfo)) (*http.Response, error) {
+	t.Helper()
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		t.Fatalf("failed to create new request: %v", err)
+	}
+
+	trace := &httptrace.ClientTrace{
+		GotConn: func(ci httptrace.GotConnInfo) {
+			t.Logf("GotConnInfo: %+v", ci)
+			f(t, ci)
+		},
+	}
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), trace))
+	return client.Do(req)
+}
+
+func shouldUseExistingConnection(t *testing.T, ci httptrace.GotConnInfo) {
+	t.Helper()
+	if !ci.Reused {
+		t.Errorf("expected an existing TCP connection to be reused, but got: %+v", ci)
+	}
+}
+
+func shouldUseNewConnection(t *testing.T, ci httptrace.GotConnInfo) {
+	t.Helper()
+	if ci.Reused {
+		t.Errorf("expected a new connection, but got: %+v", ci)
+	}
+}
+
+type neverEnding byte
+
+func (b neverEnding) Read(p []byte) (n int, err error) {
+	for i := range p {
+		p[i] = byte(b)
+	}
+	return len(p), nil
+}
+
+func closeResponseBody(t *testing.T, resp *http.Response) {
+	t.Helper()
+	if err := resp.Body.Close(); err != nil {
+		t.Errorf("unexpected error while closing the Body of the Response object: %v", err)
+	}
+}
+
+type waiter interface {
+	wait() <-chan time.Time
+	close()
+}
+
+type waitWithDuration struct {
+	after time.Duration
+}
+
+func (w waitWithDuration) wait() <-chan time.Time { return time.After(w.after) }
+func (w waitWithDuration) close()                 {}
+
+type waitWithChannelClose struct {
+	after chan time.Time
+}
+
+func (w waitWithChannelClose) wait() <-chan time.Time {
+	// for http/2, we do the following:
```

Comment:
this comment does not make sense here. This object has nothing to do with http/2.